docs/source/reference/oauth.md

```diff
@@ -0,0 +1,217 @@
+# JupyterHub and OAuth
+
+JupyterHub uses OAuth 2 internally as a mechanism for authenticating users.
+As such, JupyterHub itself always functions as an OAuth **provider**.
+More on what that means below.
+
+Additionally, JupyterHub is _often_ deployed with [oauthenticator](https://oauthenticator.readthedocs.io),
+where an external identity provider, such as GitHub or KeyCloak, is used to authenticate users.
+When this is the case, there are \*two nested
+
+This means that when you are using JupyterHub, there is always _at least one_ and often two layers of OAuth involved in a user logging in and accessing their server.
+
+Some relevant points:
+
+- Single-user servers _never_ need to communicate with or be aware of the upstream provider.
+  As far as they are concerned, only JupyterHub is an OAuth provider,
+  and how users authenticate with the Hub itself is irrelevant.
+- When talking to a single-user server,
+  there are ~always two tokens:
+  a token issued to the server itself to communicate with the Hub API,
+  and a second per-user token in the browser to represent the completed login process and authorized permissions.
+  More on this later.
+
+### Key OAuth terms
+
+- **provider** the entity responsible for managing.
+  JupyterHub is _always_ an oauth provider for JupyterHub's components.
+  When OAuthenticator is used, an external service, such as GitHub or KeyCloak, is also an oauth provider.
+- **client** An entity that requests OAuth tokens on a user's behalf.
+  JupyterHub _services_ or single-user _servers_ are OAuth clients of the JupyterHub _provider_.
+  When OAuthenticator is used, JupyterHub is itself also an OAuth _client_ for the external oauth _provider_, e.g. GitHub.
```

Comment:
`When OAuthenticator is used, JupyterHub is itself also an OAuth _client_ for the external oauth _provider_, e.g. GitHub.`: Is this really the case? For e.g. GitHub, the oauth client is the effective user (eg. @minrk), not JupyterHub.

Reply:
mmh ok, rereading makes more sense. For GitHub example, JupyterHub (the oauth client) asks GitHub 'me, JupyterHub, I ask you to give me a token so I can grab @minrk identity`. Therefor, @minrk needs to authenticate on GitHub and agree that JupyterHub has a token to query @minrk identity later on. In that sense, the client in that specific OAuth flow is indeed JupyterHub, and not @minrk

Reply:
Right. Probably worth linking to and/or cribbing from [this oauth terminology reference](https://www.oauth.com/oauth2-servers/definitions/). Users are never Clients in this language.